### PR TITLE
core, eth/downloader: fix genesis state missing due to state sync

### LINKED ISSUE #28120: v1.13.0 rewinds to block 0
#### System information

Geth version: `v1.13.0`
CL client & version: `lighthouse v4.4.1`
OS & Version: `Linux`

I restarted geth while syncing from scratch with `--state.sceme=path` and this lead (I think) the chain to start rewinding to block zero. chaindata size started to decrease. after 2 hours and another restart I decided to `removedb` (without removing ancients) and the node resumed syncing again.

Logs just before the restart:
````
INFO [09-14|09:46:16.398] IPC endpoint closed                      url=/chain/chain.ipc
WARN [09-14|09:46:16.399] Rewinding blockchain to block            target=7,944,977
INFO [09-14|09:46:24.487] Looking for peers                        peercount=0 tried=168 static=1
INFO [09-14|09:46:34.522] Looking for peers                        peercount=0 tried=116 static=1
INFO [09-14|09:46:41.793] Loaded most recent local header          number=7,944,977  hash=6ba207..e2dc1e td=10,579,629,261,200,387,371,986 age=4y3mo3w
INFO [09-14|09:46:41.793] Loaded most recent local block           number=0
        hash=d4e567..cb8fa3 td=17,179,869,184                 age=54y5mo3w
INFO [09-14|09:46:41.793] Loaded most recent local snap block      number=7,906,288  hash=623ffe..808d51 td=10,500,012,474,221,072,800,842 age=4y4mo16h
INFO [09-14|09:46:41.793] Loaded last snap-sync pivot marker       number=18,132,746
WARN [09-14|09:46:41.793] Rolled back chain segment                header=7947026->7944977 snap=7906288->7906288 block=0->0 reason="syncing canceled (requested)"
INFO [09-14|09:46:41.812] Syncing: chain download in progress      synced=43.61% chain=82.92GiB headers=7,944,977@3.14GiB bodies=7,908,044@56.50GiB receipts=7,908,044@23.28GiB eta=28h9m18.490s
INFO [09-14|09:46:41.812] Syncing: state download in progress      synced=54.81% state=152.61GiB accounts=126,154,895@24.40GiB slots=600,831,588@124.36GiB codes=616,171@3.85GiB eta=7h45m21.870s
ERROR[09-14|09:46:41.949] Beacon backfilling failed                err="content processing canceled (requested)"
INFO [09-14|09:46:41.957] Ethereum protocol stopped
INFO [09-14|09:46:41.957] Transaction pool stopped
ERROR[09-14|09:46:41.961] Failed to journal state snapshot         err="snapshot [0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544] missing"
INFO [09-14|09:46:41.961] Persisting dirty state to disk           root=d7f897..0f0544 layers=0
INFO [09-14|09:46:41.963] Persisted dirty state to disk            size=69.00B elapsed=2.252ms
INFO [09-14|09:46:41.965] Blockchain stopped
geth.service: Deactivated successfully.
Stopped geth.service - Ethereum geth client daemon by The Ethereum Foundation.
````

Logs right after the restart:
````
INFO [09-14|09:46:53.751]
INFO [09-14|09:46:53.752] Loaded most recent local header          number=7,944,977 hash=6ba207..e2dc1e td=10,579,629,261,200,387,371,986 age=4y3mo3w
INFO [09-14|09:46:53.752] Loaded most recent local block           number=0
       hash=d4e567..cb8fa3 td=17,179,869,184                 age=54y5mo3w
INFO [09-14|09:46:53.752] Loaded most recent local snap block      number=7,908,044 hash=e8f41d..7dd8dd td=10,503,796,279,546,327,487,969 age=4y4mo9h
INFO [09-14|09:46:53.752] Loaded last snap-sync pivot marker       number=18,132,746
ERROR[09-14|09:46:53.752] Unexpected trie node in disk             owner=000000..000000 path=[] expect=d7f897..0f0544 got=5d7a4d..b04697
WARN [09-14|09:46:53.800] Head state missing, repairing            number=0
        hash=d4e567..cb8fa3
ERROR[09-14|09:46:53.800] Unexpected trie node in disk             owner=000000..000000 path=[] expect=d7f897..0f0544 got=5d7a4d..b04697
ERROR[09-14|09:46:53.800] Unexpected trie node in disk             owner=000000..000000 path=[] expect=d7f897..0f0544 got=5d7a4d..b04697
INFO [09-14|09:46:54.058] Opened ancient database                  database=/chain/geth/geth/chaindata/ancient/state readonly=false
INFO [09-14|09:46:54.058] Rebuilt trie database                    root=56e81f..63b421
INFO [09-14|09:46:54.627] Loaded most recent local header          number=7,944,977  hash=6ba207..e2dc1e td=10,579,629,261,200,387,371,986 age=4y3mo3w
INFO [09-14|09:46:54.627] Loaded most recent local block           number=0
        hash=d4e567..cb8fa3 td=17,179,869,184                 age=54y5mo3w
INFO [09-14|09:46:54.627] Loaded last snap-sync pivot marker       number=18,132,746
ERROR[09-14|09:46:54.627] Truncating ancient chain                 from=7,944,977 to=0
WARN [09-14|09:46:54.628] Rewinding blockchain to block            target=0
````

Logs after 2 hours where I attempted another restart (at this point chaindata was almost half the size, ancients decreased in size too)
````
INFO [09-14|10:36:32.740] Loaded most recent local header          number=7,664,518 hash=dd5b9a..9c5985 td=10,011,359,787,843,088,592,518 age=4y5mo1w
INFO [09-14|10:36:32.740] Loaded most recent local block           number=0
       hash=d4e567..cb8fa3 td=17,179,869,184                 age=54y5mo3w
INFO [09-14|10:36:32.740] Loaded last snap-sync pivot marker       number=18,132,746
ERROR[09-14|10:36:32.740] Truncating ancient chain                 from=7,664,518 to=0
WARN [09-14|10:36:32.740] Rewinding blockchain to block            target=0
WARN [09-14|10:36:32.751] Truncating freezer table                 database=/chain/geth/geth/chaindata/ancient/chain table=diffs items=7,664,520 limit=7,664,518
````

### PR DESCRIPTION
**Pull request by rjl493456442**
Fixes #28120 

This pull request fixes two main flaws introduced by path-based scheme. 

##

a) In the blockchain repair stage, initiate the repair process to restore the missing genesis state if the head block is still the genesis block.

This is indeed a unique and exceptional scenario that can only occur in the path-based scheme. The root cause of this issue is the state syncer overwriting the persistent genesis state, leading to a situation where no complete and usable state is available. 

This pull requests fixes this issue by restoring the genesis state. It won't affect the correctness of following state sync, due to the fact that the written genesis state is **complete** and state healer will address and fix any inconsistencies between the genesis state and sync target.

It's an necessary operation to guarantee the associated state of head block is present and usable, otherwise it will break the assumption in many places which blindly create the state of head block(e.g. txpool).

##

b) Disable trie database and state snap whenever state sync is launched.

State sync overwrites the persistent state, and this action can disrupt states that reference the current persistent state. The appropriate solution in such a situation is to mark the entire trie database as unusable, which prevents any subsequent state reads, and raises a `layer stale` error if the read operation is done by the prev-created **stale** state. The trie database will remain in this state until the state sync process is completed, and a usable persistent state is assembled.

Besides, these two components will be disabled whenever a new state sync cycle is launched, not just at the beginning of the whole sync process. It's because the complete state will be re-synced because of pivot point moving.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -340,28 +340,38 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis
...
-			if snapDisk != 0 {
-				rawdb.WriteSnapshotRecoveryNumber(bc.db, snapDisk)
-			}
+		if head.Number.Uint64() == 0 {
+			// The genesis state is missing, which is only possible in the path-based
+			// scheme. This situation occurs when the state syncer overwrites it.
+			//
+			// The solution is to reset the state to the genesis state. Although it may not
+			// match the sync target, the state healer will later address and correct any
+			// inconsistencies.
```
Comment (jsvisa) [inline core/blockchain.go:349]: I hope to add a log here as well

File: eth/downloader/downloader.go [hunk 2]
```diff
@@ -398,7 +398,14 @@ func (d *Downloader) synchronise(id string, hash common.Hash, td, ttd *big.Int,
 		log.Info("Block synchronisation started")
 	}
 	if mode == SnapSync {
-		// Snap sync uses the snapshot namespace to store potentially flakey data until
+		// Snap sync will directly modify the persistent state, making the entire
+		// trie database unusable until the state is fully synced. To prevent any
+		// subsequent state reads, explicitly disable the trie database and state
+		// syncer is responsible to address and correct any state missing.
+		if d.blockchain.TrieDB().Scheme() == rawdb.PathScheme {
```
Comment (rjl493456442) [inline eth/downloader/downloader.go:405]: The trie database and state snapshot are explicitly disabled here, at the beginning of the **snap sync** procedure.

However, this approach may need reconsideration. During a snap sync procedure, we may initiate multiple state sync cycles with different targets, particularly when the pivot point shifts. It is possible that a state has already been sync'd, and we begin synchronizing another one.

The enabling and disabling of these components should align more cohesively.

Comment (rjl493456442) [reply eth/downloader/downloader.go:405]: With the fix #28126 , we have the guarantee that the committed state won't be resync'd. In another word, for a snap sync process, we either end up with a complete state and re-enable the trie database and state snapshot for **single time** or exits abnormally. It's safe to invalidate the trie database and state snapshot only at the beginning of snap sync.

**Comment by rjl493456442** [comment]
A few things to highlight for reviewers, the trie database and state snapshot will be disabled at the beginning of snap sync, and will only be re-enabled if the state is completely synced. 

Therefore, If restart occurs before the sync is finished, the **trie database will be in the uninitialized status**, and results in the re-init genesis operation here https://github.com/ethereum/go-ethereum/blob/master/core/genesis.go#L335.

But it's correct, re-init genesis state in the middle of snap sync **won't affect the correctness**.

File: core/blockchain.go (hunk 1 above)
Comment (rjl493456442) [reply core/blockchain.go:349]: `resetState ` will print something. But one thing to mention that by disabling the trie database at the beginning of snap sync, the trie database will be in "uninitialized" status until the state is fully sync'd. 

In the next restart, the missing genesis status will be re-inited in `SetupGenesisBlockWithOverride`. Therefore, this logic is a bit redundant, but still necessary.

**Review by karalabe — APPROVED** [review]
SGTM

**Pull request merged**